component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java

```diff
@@ -332,26 +339,74 @@ public int getSize() throws Exception {
+    queryStringBuilder.append(" AND identity_1.enabled = ").append(dbBoolTrue).append(" \n");
+
+    return getEntityManager().createNativeQuery(queryStringBuilder.toString());
+  }
+
+
   private Query getIdentitiesQuerySortedByField(String providerId,
                                                 String firstCharacterFieldName,
                                                 char firstCharacter,
                                                 String sortField,
-                                                String sortDirection) {
+                                                String sortDirection,
+                                                boolean excludeExternal) {
     // Oracle and MSSQL support only 1/0 for boolean, Postgresql supports only
     // TRUE/FALSE, MySQL supports both
     String dbBoolFalse = isOrcaleDialect() || isMSSQLDialect() ? "0" : "FALSE";
     String dbBoolTrue = isOrcaleDialect() || isMSSQLDialect() ? "1" : "TRUE";
     // Oracle Dialect in Hibernate 4 is not registering NVARCHAR correctly, see
     // HHH-10495
     StringBuilder queryStringBuilder = null;
-    if (isOrcaleDialect()) {
-      queryStringBuilder = new StringBuilder("SELECT to_char(identity_1.remote_id), identity_1.identity_id \n");
-    } else if (isMSSQLDialect()) {
-      queryStringBuilder = new StringBuilder("SELECT try_convert(varchar(200), identity_1.remote_id) as remote_id , identity_1.identity_id, try_convert(varchar(200) \n");
+    if (excludeExternal) {
+      if (isOrcaleDialect()) {
+        queryStringBuilder = new StringBuilder("SELECT DISTINCT to_char(identity_1.remote_id), identity_1.identity_id, identity_prop.value \n");
+      } else if (isMSSQLDialect()) {
+        queryStringBuilder = new StringBuilder("SELECT DISTINCT try_convert(varchar(200), identity_1.remote_id) as remote_id , identity_1.identity_id, identity_prop.value, try_convert(varchar(200) \n");
+      } else {
+        queryStringBuilder = new StringBuilder("SELECT DISTINCT identity_1.remote_id, identity_1.identity_id, identity_prop.value \n");
+      }
+      queryStringBuilder.append(" FROM SOC_IDENTITIES identity_1 \n");
+      queryStringBuilder.append(" INNER JOIN SOC_IDENTITY_PROPERTIES identity_prop_1 \n");
+      queryStringBuilder.append("   ON identity_1.identity_id = identity_prop_1.identity_id \n");
+      queryStringBuilder.append("   AND NOT EXISTS ( SELECT properties_tmp.identity_id FROM SOC_IDENTITY_PROPERTIES as properties_tmp \n");
+      queryStringBuilder.append("   WHERE properties_tmp.identity_id = identity_1.identity_id \n");
+      queryStringBuilder.append("   AND properties_tmp.name = 'external' \n");
+      queryStringBuilder.append("   AND properties_tmp.value = ").append(dbBoolTrue).append(") \n");
     } else {
-      queryStringBuilder = new StringBuilder("SELECT identity_1.remote_id, identity_1.identity_id \n");
+      if (isOrcaleDialect()) {
```

Comment:
All this block can be put outside the if/else since it not changes

Reply:
Its changed with DISTINCT